components-api/src/main/scala/pl/touk/nussknacker/engine/api/typed/AssignabilityDeterminer.scala

```diff
@@ -0,0 +1,251 @@
+package pl.touk.nussknacker.engine.api.typed
+
+import cats.data.Validated._
+import cats.data.{NonEmptyList, Validated, ValidatedNel}
+import cats.implicits.{catsSyntaxValidatedId, _}
+import org.apache.commons.lang3.ClassUtils
+import pl.touk.nussknacker.engine.api.typed.typing._
+
+/**
+  * This class determine whether we can convert a class to another one. We provide two modes of conversion -
+ * 1. Loose conversion is based on the fact that TypingResults are
+ * sets of possible supertypes with some additional restrictions (like TypedObjectTypingResult). It is basically how SpEL
+ * can convert things. Like CommonSupertypeFinder it's in the spirit of "Be type safe as much as possible, but also provide some helpful
+ * conversion for types not in the same jvm class hierarchy like boxed Integer to boxed Long and so on".
+ * 2. Strict conversion checks whether we can convert the type without losing precision. Eg only widening numerical types
```

Comment:
So I dug into this and its more complex unfortunately, but it doesn't matter for us, we just have to temper our expectations for how strict this strict check is :smile: 

Take a look at: https://docs.oracle.com/javase/specs/jls/se11/html/jls-5.html 5.1.2. Widening Primitive Conversion

Widening conversions are what we have in our strict check, just extended with BigInteger and BigDecimal. Thats what we have right now in the strict check.

What we don't have is check if the conversion can happen without loss of precision.

"A widening primitive conversion from float to double that is not strictfp may lose information about the overall magnitude of the converted value.

A widening primitive conversion from int to float, or from long to float, or from long to double, may result in loss of precision - that is, the result may lose some of the least significant bits of the value. In this case, the resulting floating-point value will be a correctly rounded version of the integer value, using IEEE 754 round-to-nearest mode ([§4.2.4](https://docs.oracle.com/javase/specs/jls/se11/html/jls-4.html#jls-4.2.4))."

If we allow for example int -> float, we can't say we have conversions without losing precision.

Intuitively, I'd say we should allow only precision-loss-less conversion in a method thats called `isAssignableStrict`, but `isAssignable` from ClassUtils also doesnt take that into account, so its ok for me either way, its up to you. Just if we do it with allowing precision loss, then lets remove the `without losing precision` - just say its only widening conversions.

Reply:
I'd say since now it's not needed i will leave the code as it is and just reword the comment. Making it without loss of precision would require adding a lot of logic probably. I changed the comment.